### PR TITLE
Hackathon 2023-02-21

### PR DESCRIPTION
**Pull request by alimanfoo**
Opening this PR to discuss work in the hackathon-21-02-23 branch.

### CODE REVIEW

**Comment by alimanfoo** [comment]
Hi @sanjaynagi, I've opened this PR for our current development branch, just as a place to communicate any changes we make or other discussion.

I'm fine to leave this branch open and keep working on it together as we have been for a while, might be convenient rather than making lots of PRs.

**Comment by alimanfoo** [comment]
Btw just pushed a commit to rename params.yaml to config.yaml. It was bothering me because we also use "params" to talk about the parameters in a notebook than can be set when run via papermill, so to avoid confusion I thought I'd rename.

**Comment by alimanfoo** [comment]
I've started prototyping the h12 signal detection notebook, WIP.

**Comment by alimanfoo** [comment]
OK, H12 signal detection should be working. The notebook h12-signal-detection runs for a given cohort and contig, and fits exponential peaks to the H12 values, writing the signals discovered to a CSV file at `build/h12-signal-detection/{cohort}_{contig}.csv`.

**Comment by sanjaynagi** [comment]
That's awesome!

Its fully integrated with the snakemake workflow now and working on the AG1000G-BF-A example.

**Comment by alimanfoo** [comment]
Super cool.

**Comment by sanjaynagi** [comment]
Have added a WIP of the generate-toc notebook. Dont think the output is perfect yet.

Also added a calculation of a cohorts mean latitude and longitude to the setup-cohorts notebook, so that I can start developing the mapping functions. We may want to change the use of mean later to something more appropriate, but the small geographic scale of each cohort should make this OK for now.

**Comment by sanjaynagi** [comment]
Just added a github actions workflow (running the pipeline on Ag1000G-BF-A). 

For output and input files, I've had to remove the use of `here()` from pyprojroot though, as it meant output files were being generated in the `root/build/` directory when during testing/CI they should be under `root/.tests/build`. I think the workflow works fine without `here()`, that's really just useful if we are actively developing a notebook.

**Comment by sanjaynagi** [comment]
Changed my mind and decided to continue to use `here()` in the notebooks and instead run the github actions in the root directory. I guess It will help while developing. 

This meant needing to be able to pass a different config.yaml so, instead of reading the config at `workflow/config.yaml' in each notebook, I now directly pass the parameters directly as a yaml file with papermill. This is more snakemake-like, and means we can override the config at the command line with --configfile.

**Comment by alimanfoo** [comment]
> This meant needing to be able to pass a different config.yaml so, instead of reading the config at `workflow/config.yaml' in each notebook, I now directly pass the parameters directly as a yaml file with papermill. This is more snakemake-like, and means we can override the config at the command line with --configfile.

Thanks @sanjaynagi, just to say this sounds good to me.

**Comment by alimanfoo** [comment]
Hi @sanjaynagi, just to let you know I've pushed commits which add a file `workflow/notebooks/geolocate-cohorts.ipynb`. This notebook reads in the `build/cohorts.csv` file, reads in data from geoboundaries, and makes the join using admin2 unit names, with some hacks to deal with name mismatches and boundary changes. 

The notebook outputs a new file `build/cohorts.geojson` which includes columns "representative_lat" and "representative_lon" which provide coordinates for a representative point for the corresponding admin2 unit.

You should be able to read this in with code like:

```python
import geopandas as gpd
gdf_cohorts = gpd.read_file(here() / "build" / "cohorts.geojson")
```

(Note this file also includes geometry providing the simplified boundaries of the admin2 units, but as discussed we'll probably not be able to use that because we can't attach a popup to polygons.)

**Comment by sanjaynagi** [comment]
I've now made draft notebooks for all the site pages, they're all WIP (especially the cohorts and country pages), but they're all incorporated into the snakemake workflow now and seem to be working correctly. Will start to finalise the site pages and then will ask you for edits. 

We'll need to write a config for the jupyter-book, will make a start on that too.

**Comment by sanjaynagi** [comment]
Hey @alimanfoo. Do you know of any way to hide all the cell inputs in the jupyter book? All I can see at the minute is having to do it cell by cell (!) or some annoying looking python scripting. Surely there is just a simple flag in the config!

https://jupyterbook.org/en/stable/interactive/hiding.html

https://jupyterbook.org/en/stable/content/metadata.html?highlight=python#add-tags-using-python-code

**Comment by sanjaynagi** [comment]
Ignore the previous message, there were very few overall cells to add tags to, did it manually.

**Comment by sanjaynagi** [comment]
Also, we didnt need the papermill set kernel rule - removed.

**Comment by alimanfoo** [comment]
😂

**Comment by alimanfoo** [comment]
> Hey @alimanfoo. Do you know of any way to hide all the cell inputs in the jupyter book? All I can see at the minute is having to do it cell by cell (!) or some annoying looking python scripting. Surely there is just a simple flag in the config!

Unfortunately doesn't seem possible yet, issue still open here... https://github.com/executablebooks/jupyter-book/issues/666

Shame as nbconvert has an option to remove all inputs.

**Comment by alimanfoo** [comment]
Sorry accidentally closed while trying to type on mobile phone with my left thumb, doh.

**Comment by sanjaynagi** [comment]
FYI - one problem I'm having at the minute is that Jupyter-book does not seem to recognise the glue function as possible title for a notebook - I.e - on the left sidebar, all the country pages just say 'country page' rather than 'Mali', 'Burkina Faso' etc. 

The only thing I can think of is I may have to write a script to programmatically add the titles!

**Comment by sanjaynagi** [comment]
So I've written a notebook `add-headers.ipynb` which adds appropriate headers to all of the notebooks in one go, that works fine. However, snakemakes handling of files and the checkpoint system is becoming a pain. it's the first time I've thought that I bet nextflow can handle this case. 

Because I need the final `jupyter-book build` rule to take all the appropriate notebooks as input, I need another rule which specifies them as output (otherwise snakemake will complain they cant be generated). The `add-headers.ipynb` notebook should do this, however, we are not allowed to use functions in the `output` of a rule, only the input. Because we dont know the list of countries at the start of the workflow, I cant do something like

output: expand("docs/country/{country}.ipynb", country=country_list)

I need to get countries by using a function which has checkpoints.geolocate_cohorts.get().output.cohorts_geojson within it. The only way around it, would be to have three add-header rules, one for each type of notebook (country, cohort, chromosome) and use wildcards, then the output could simply be

output: "docs/country/{country}.ipynb"

and wildcards would sort the rest.

I guess its not that complicated to make three rules, but its an annoyance which seems unnecessary.

**Comment by alimanfoo** [comment]
> FYI - one problem I'm having at the minute is that Jupyter-book does not seem to recognise the `glue` function as possible title for a notebook - I.e - on the left sidebar, all the country pages just say 'country page' rather than 'Mali', 'Burkina Faso' etc.

That is annoying.

File: workflow/rules/site.smk [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+    output:
+        nb = "build/notebooks/add_headers/{contig}.ipynb",
+        chrom_nb = "docs/genome/ag-{contig}.ipynb",
+    log:
+        "logs/add_headers/{contig}.log"
+    conda:
+        f"{workflow.basedir}/../environment.yml"
+    shell:
+        """
+        papermill {input.nb} {output.nb} -k selection-atlas -p input_nb {input.chrom_nb} -p output_nb {output.chrom_nb} -p wildcard {wildcards.contig} -p type chrom 2> {log}
```
Comment (alimanfoo) [inline workflow/rules/site.smk:98]: This is fun, using a notebook to modify a notebook 🙂

File: workflow/notebooks/add-headers.ipynb [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    "def add_header_and_remove_tag(input_nb, output_nb, header, cell=3):\n",
+    "\n",
+    "    with open(input_nb, mode= \"r\", encoding= \"utf-8\") as f:\n",
+    "        my_nb = json.loads(f.read())\n",
+    "\n",
+    "    my_nb['cells'][cell]['source'] = header\n",
+    "    my_nb['cells'][1]['metadata']['tags'] = ['remove-input']\n",
+    "    \n",
+    "    with open(output_nb, 'w') as f:\n",
+    "        json.dump(my_nb, f)"
```
Comment (alimanfoo) [inline workflow/notebooks/add-headers.ipynb:45]: Nice hacking 🙂

**Comment by alimanfoo** [comment]
Nice one getting around the hurdles with snakemake and jupyterbook.

**Comment by sanjaynagi** [comment]
No idea whats going on here, if you get a chance at some point could you try and build the site, and see what happens on your side?

**Comment by alimanfoo** [comment]
> The error is something about an unknown MIME type? "WARNING: skipping unknown output mime type: application/vnd.bokehjs_exec.v0+json [mystnb.unknown_mime_type]"
> 
> EDIT - worth adding that bokeh plots are working in other notebooks. Will figure it out.

No idea what's happening there. Only wondering if bokeh output_notebook() has something to do with it.

**Comment by sanjaynagi** [comment]
> 
> No idea what's happening there. Only wondering if bokeh output_notebook() has something to do with it.

yessss @alimanfoo  `output-notebook()` worked!!! praise the lord!
Still strange that it's working in other notebooks without `output-notebook()` though. No idea. 

The CI is finally back working again too. Btw, it turned out that we did need the rule that set the kernel, and we did need the `-k selection-atlas` argument to papermill. 

awesome. So I think we have a functioning draft prototype of the site build now? There is still some tidying to do of each notebook (mainly the text, I just copied what you put in figma mostly), and we also need to add iHS and G123. Now seems like a good time to merge this PR and start working normally, what do you think? :)

**Comment by alimanfoo** [comment]
> Now seems like a good time to merge this PR and start working normally, what do you think? :)

Yeah sounds great, go for it!

**Comment by alimanfoo** [comment]
> yessss @alimanfoo `output-notebook()` worked!!! praise the lord!
> Still strange that it's working in other notebooks without `output-notebook()` though. No idea.

Awesome. Don't know either, possibly something to do with the fact that malariagen_data calls output_notebook() internally under some circumstances - it's such a common thing for users to forget to do then wonder why plots aren't appearing, so we added it as part of `Ag3.__init__()` IIRC.

**Comment by sanjaynagi** [comment]
> 
> Awesome. Don't know either, possibly something to do with the fact that malariagen_data calls output_notebook() internally under some circumstances - it's such a common thing for users to forget to do then wonder why plots aren't appearing, so we added it as part of `Ag3.__init__()` IIRC.

Indeed, its happened to me before, but because it was working in the other NB I was totally baffled!

>Yeah sounds great, go for it!

Awesome, merging!

**Pull request closed without merging**